### PR TITLE
Closes: Fieldset validator for releated fields

### PR DESCRIPTION
**Pull request by dkorbel**
**What this PR does / why we need it**:
This feature adds additional validator that work for multiply fields.

This is a new validator that works for a set of fields that are related to each other. Based on the value of one field, it can validate a subset of fields for that value.

For example, in a purchase contact information form, we need different values for a company and a person. So, we add a 'select' field where we can choose whose details we need: either a person or a company. Then, we add fields for both individuals and companies. The validator will only validate the fields indicated by the first 'select' field.

**Which issue(s) this PR closes**:
It comes from Fairbirds team.

Example usage:
#datasetField	name	title	validation
	responsible_party	Responsible party	"[{""name"":""standard_input""}]"
	responsible_party_type	Type	"[{""name"":""standard_input""}]"
	responsible_party_person_first_name	First name	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""person""}}]"
	responsible_party_person_surnames	Surname(s)	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""person""}}]"
	responsible_party_organization_name	Organization name	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""organization""}}]"
	responsible_party_organization_address_city	City	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""organization""}}]"
	responsible_party_organization_address_postal-code	Postal code	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""organization""}}]"
	responsible_party_organization_address_administrative-area	Administrative area	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""organization""}}]"
	responsible_party_organization_address_country	Country	"[{""name"":""required_by_value_dependant"",""parameters"":{""context"":[""DATASET""], ""runOnEmpty"":""true"", ""dependantField"": ""responsible_party_type"", ""dependantFieldValue"": ""organization""}}]"
	responsible_party_email_address	Email address	"[{""name"":""standard_input""}]"
	responsible_party_orcid	ORCID	"[{""name"":""standard_input""}]"
#controlledVocabulary	DatasetField	Value	identifier
	responsible_party_type	Person	person
	responsible_party_type	Organization	organization

[releatedFieldValidator.csv](https://github.com/user-attachments/files/16992492/releatedFieldValidator.csv)

### CODE REVIEW

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+import java.util.HashMap;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RequiredByValueDependantFieldValidatorTest {
```
Comment (sniewczas) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidatorTest.java:28]: The build is failing.

```
[ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.269 s <<< FAILURE! - in edu.harvard.iq.dataverse.validation.field.validators.RequiredByValueDependantFieldValidatorTest
[ERROR] edu.harvard.iq.dataverse.validation.field.validators.RequiredByValueDependantFieldValidatorTest  Time elapsed: 0.269 s  <<< ERROR!
org.mockito.exceptions.misusing.NotAMockException: Argument passed to Mockito.mockingDetails() should be a mock, but is an instance of class java.lang.Class!
```

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredDependantFieldValidator.java [hunk 2]
```diff
@@ -10,6 +10,24 @@
...
+ * To use this validator you need to set up `dependantField` parameter in your DatasetFieldType validator configuration
+ * <pre>
+ * F.e. `{"name":"required_dependant","parameters":{"context":["DATASET"], "dependantField": "OriginTitle"}}`
+ * F.e. `{"name":"required_dependant","parameters":{"context":["DATASET"], "dependantField": "OtherTitle"}}`
+ * </pre>
+ * @see edu.harvard.iq.dataverse.persistence.dataset.ValidatableField
+ * @see edu.harvard.iq.dataverse.persistence.dataset.DatasetField
+ * @see edu.harvard.iq.dataverse.persistence.dataset.DatasetFieldType#getValidation()
+ *
+ * @author Daniel Korbel, Krzysztof Mądry, Sylwester Niewczas
```
Comment (sniewczas) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredDependantFieldValidator.java]: I don't think attributing authorship to us here is appropriate. Let's at least mention the original author first.

**Review by sniewczas — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    // -------------------- LOGIC --------------------
+
+    @Override
+    public String getName() {
+        return "required_by_value_dependant";
+    }
+
+    @Override
+    protected FieldValidationResult validateWithDependantField(ValidatableField field, ValidatableField dependantField, Map<String, Object> params, Map<String, ? extends List<? extends ValidatableField>> fieldIndex) {
+        if (dependantField.getValidatableValues().stream().anyMatch(actualValue -> StringUtils.equalsIgnoreCase(actualValue, (String) params.get(DEPENDANT_FIELD_VALUE_PARAM)))) {
```
Comment (sniewczas) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidator.java]: Nitpick: These super-long lines could use a line break. Let's not force people to scroll vertically.

**Review by madryk — APPROVED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredDependantFieldValidator.java (hunk 2 above)
Comment (dkorbel) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredDependantFieldValidator.java]: Sure

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidator.java (hunk 3 above)
Comment (dkorbel) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/validation/field/validators/RequiredByValueDependantFieldValidator.java]: Sure

**Review by sniewczas — APPROVED** [review]

**Pull request merged**